### PR TITLE
DM: adjust loader's dir to be add subdirectory

### LINKED ISSUE #4360: Support S3 for DM dump and load
### Is your feature request related to a problem?

Upstream data volumes can be large and variable, making them difficult and costly to manage and plan for using EBS. So users want to use DM on s3 storage.

### Describe the feature you'd like

DM supports s3 storage to dump and load data files.

### Describe alternatives you've considered

_No response_

### Teachability, Documentation, Adoption, Migration Strategy

_No response_

### PR DESCRIPTION
**Pull request by WizardXiao**
### What problem does this PR solve?

Issue Number: ref #4360

### What is changed and how it works?
Adjust path to be subdirectory instead of suffix when we adjust loader's dir in subtask. 

### Check List 

Tests 

 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/dm/config/subtask.go [hunk 1]
```diff
@@ -415,7 +416,7 @@ func (c *SubTaskConfig) Adjust(verifyDecryptPassword bool) error {
 		} else {
 			dirSuffix = c.Name
 		}
-		newDir, err := storage.AdjustPath(c.LoaderConfig.Dir, "."+dirSuffix)
+		newDir, err := storage.AdjustPath(c.LoaderConfig.Dir, string(filepath.Separator)+dirSuffix)
```
Comment (WizardXiao) [inline dm/dm/config/subtask.go:419]: `storage.AdjustPath` has no change and may be we can use in future, not only add `/xxx`.

**Review by lance6716 — APPROVED** [review]
will upgrading from an old version get affected?

**Review by lichunzhu — APPROVED** [review]

**Comment by csuzhangxc** [comment]
> will upgrading from an old version get affected?

Can it work if only part of dump data has been imported during the upgrade of DM?

**Comment by WizardXiao** [comment]
> > will upgrading from an old version get affected?
> 
> Can it work if only part of dump data has been imported during the upgrade of DM?

no, it can't, it need full stage has done and syncer is runing now.

**Comment by WizardXiao** [comment]
After discuss with @sunzhaoyang , we decide to just change s3 first, because change local path may have
 risk of compatibility.

File: dm/dm/config/subtask.go [hunk 2]
```diff
@@ -411,11 +411,11 @@ func (c *SubTaskConfig) Adjust(verifyDecryptPassword bool) error {
 		// add suffix
 		var dirSuffix string
 		if isS3 {
-			dirSuffix = c.Name + "." + c.SourceID
+			dirSuffix = "/" + c.Name + "." + c.SourceID
```
Comment (Ehco1996) [inline dm/dm/config/subtask.go:414]: please add comment for this change

**Review by Ehco1996 — APPROVED** [review]
LGTM

File: dm/dm/config/subtask.go (hunk 2 above)
Comment (Ehco1996) [reply dm/dm/config/subtask.go:415]: i suddenly find that we shoud adapt this change when convert subtask to OpenAPITask
see here https://github.com/pingcap/tiflow/blob/50c98a6ac9eb456ac1cd47d871349b92976eb868/dm/dm/config/task_converters.go#L358

Comment (WizardXiao) [reply dm/dm/config/subtask.go:415]: i add trim path suffix at [this commit](https://github.com/pingcap/tiflow/pull/4897/commits/b481c0f6b746688b12d48798538bc9e49ea34c52), please take a look.

File: dm/dm/config/task_converters.go [hunk 3]
```diff
@@ -355,9 +357,21 @@ func SubTaskConfigsToTaskConfig(stCfgs ...*SubTaskConfig) *TaskConfig {
...
+			dirSuffix = "/" + c.Name + "." + stCfg.SourceID
+		} else {
+			// TODO we will dump local file to dir's subdirectory, but it may have risk of compatibility, we will fix in other pr
+			dirSuffix = "." + c.Name
+		}
 		// if ends with the task name, we remove to get user input dir.
-		loaderCfg.Dir = strings.TrimSuffix(loaderCfg.Dir, dirSuffix)
+		loaderCfg.Dir, err = storage.TrimPath(loaderCfg.Dir, dirSuffix)
+		// because dir comes form subtask, there should not have error.
+		log.L().Warn("parse config comes from subtask error.", zap.Error(err))
```
Comment (Ehco1996) [inline dm/dm/config/task_converters.go:373]: if err!=nil{
}

File: dm/dm/config/task_converters.go [hunk 4]
```diff
@@ -445,9 +459,20 @@ func SubTaskConfigsToOpenAPITask(subTaskConfigList []*SubTaskConfig) *openapi.Ta
...
+		dirSuffix = "/" + oneSubtaskConfig.Name + "." + oneSubtaskConfig.SourceID
+	} else {
+		// TODO we will dump local file to dir's subdirectory, but it may have risk of compatibility, we will fix in other pr
+		dirSuffix = "." + oneSubtaskConfig.Name
+	}
 	// if ends with the task name, we remove to get user input dir.
-	oneSubtaskConfig.LoaderConfig.Dir = strings.TrimSuffix(oneSubtaskConfig.LoaderConfig.Dir, dirSuffix)
+	oneSubtaskConfig.LoaderConfig.Dir, err = storage.TrimPath(oneSubtaskConfig.LoaderConfig.Dir, dirSuffix)
+	// because dir comes form subtask, there should not have error.
+	log.L().Warn("parse config comes from subtask error.", zap.Error(err))
```
Comment (Ehco1996) [inline dm/dm/config/task_converters.go:474]: ditto

File: dm/dm/config/task_converters.go (hunk 3 above)
Comment (WizardXiao) [reply dm/dm/config/task_converters.go]: ok

File: dm/dm/config/task_converters.go (hunk 4 above)
Comment (WizardXiao) [reply dm/dm/config/task_converters.go]: ok

File: dm/pkg/storage/utils.go [hunk 5]
```diff
@@ -40,21 +41,43 @@ func AdjustPath(rawURL string, uniqueID string) (string, error) {
...
+	if rawURL == "" || uniqueID == "" {
+		return rawURL, nil
+	}
+	u, err := bstorage.ParseRawURL(rawURL)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	// not url format, we don't use url library to avoid being escaped or unescaped
+	if u.Scheme == "" {
+		// avoid duplicate add uniqueID, and trim suffix '/' like './dump_data/'
```
Comment (lance6716) [inline dm/pkg/storage/utils.go:72]: Seems this function is only used for the subtask config, and subtask config will not have `/` suffix? we can remove this comment or in later PR

**Review by lance6716 — COMMENTED** [review]
lgtm

File: dm/pkg/storage/utils.go (hunk 5 above)
Comment (WizardXiao) [reply dm/pkg/storage/utils.go:72]: This is for handle the last `/` of user configuration like `./dump_data/`.  this issue was found out by @sunzhaoyang when discuss how to add subdir.

Comment (WizardXiao) [reply dm/pkg/storage/utils.go:72]: ok, i will fix this comment in later pr.

**Pull request merged**